tests/v2/validation/upgrade/kubernetes.go

```diff
@@ -26,23 +33,127 @@ const (
+	clusterMeta, err := clusters.NewClusterMeta(client, clusterName)
+	require.NoError(t, err)
+	require.NotNilf(t, clusterMeta, "Couldn't get the cluster meta")
+	logrus.Infof("[%v]: Provider is: %v, Hosted: %v, Imported: %v , Local: %v", clusterName, clusterMeta.Provider, clusterMeta.IsHosted, clusterMeta.IsImported, clusterMeta.IsLocal)
+
+	initCluster, err := bundledclusters.NewWithClusterMeta(clusterMeta)
+	require.NoError(t, err)
+
+	cluster, err := initCluster.Get(client)
+	require.NoError(t, err)
+
+	versions, err := cluster.ListAvailableVersions(client)
+	require.NoError(t, err)
+	logrus.Infof("[%v]: Available versions for the cluster: %v", clusterName, versions)
+
+	version := getVersion(t, clusterName, versions, isLatestVersion, versionToUpgrade)
+	require.NotNilf(t, version, "Couldn't get the version")
+	logrus.Infof("[%v]: Selected version: %v", clusterName, *version)
+
+	updatedCluster, err := cluster.UpdateKubernetesVersion(client, version)
+	require.NoError(t, err)
+
+	logrus.Infof("[%v]: Validating sent update request for kubernetes version of the cluster", clusterName)
+	validateKubernetesVersions(t, client, updatedCluster, version, isCheckingCurrentCluster)
+
+	logrus.Infof("[%v]: Waiting cluster to be upgraded and ready", clusterName)
+	err = clusters.WaitClusterToBeUpgraded(client, clusterMeta.ID)
+	require.NoError(t, err)
+
+	if clusterMeta.IsHosted {
+		updatedCluster.UpdateNodepoolKubernetesVersions(client, version)
+
+		logrus.Infof("[%v]: Validating sent update request for nodepools kubernetes versions of the cluster", clusterName)
+		validateNodepoolVersions(t, client, updatedCluster, version, isCheckingCurrentCluster)
+
+		err = clusters.WaitClusterToBeUpgraded(client, clusterMeta.ID)
+		require.NoError(t, err)
+
+		logrus.Infof("[%v]: Validating updated cluster's nodepools kubernetes versions", clusterName)
+		validateNodepoolVersions(t, client, updatedCluster, version, !isCheckingCurrentCluster)
+	}
+
+	if strings.Contains(versionToUpgrade, RKE1KubeVersionCheck) {
+		err = nodestat.AllManagementNodeReady(client, clusterMeta.ID, defaults.ThirtyMinuteTimeout)
+	} else if strings.Contains(versionToUpgrade, RKE2KubeVersionCheck) || strings.Contains(versionToUpgrade, K3SKubeVersionCheck) {
+		err = nodestat.AllMachineReady(client, clusterMeta.ID, defaults.TenMinuteTimeout)
+	}
+	require.NoError(t, err)
+
+	clusterToken, err := clusters.CheckServiceAccountTokenSecret(client, clusterName)
+	require.NoError(t, err)
+	assert.NotEmpty(t, clusterToken)
+
+	if psact == string(provisioninginput.RancherPrivileged) || psact == string(provisioninginput.RancherRestricted) || psact == string(provisioninginput.RancherBaseline) {
+		err := psadeploy.CreateNginxDeployment(client, clusterMeta.ID, psact)
+		require.NoError(t, err)
+	}
+
+	podErrors := pods.StatusPods(client, clusterMeta.ID)
+	assert.Empty(t, podErrors)
```

Comment:
imo all of this should be pulled out into its own `verify` function.

Reply:
Similar to above comment.

Reply:
Resolved